File: 0239_sliding_window_maximum/monotonic_deque/python/solution.py
from collections import deque


def max_sliding_window(nums: list[int], k: int) -> list[int]:
    dq: deque[int] = deque()
    result: list[int] = []

    for i, value in enumerate(nums):
        while dq and dq[0] <= i - k:
            dq.popleft()

        while dq and nums[dq[-1]] <= value:
            dq.pop()

        dq.append(i)

        if i >= k - 1:
            result.append(nums[dq[0]])

    return result

File: 0239_sliding_window_maximum/monotonic_deque/clojure/src/solution.clj
(ns solution)

(defn pop-expired [dq left-bound]
  (if (and (seq dq) (<= (first dq) left-bound))
    (subvec dq 1)
    dq))

(defn pop-smaller-from-back [dq nums value]
  (loop [current dq]
    (if (and (seq current)
             (<= (nth nums (peek current)) value))
      (recur (pop current))
      current)))

(defn max-sliding-window [nums k]
  (loop [i 0
         dq []
         result []]
    (if (= i (count nums))
      result
      (let [dq1 (pop-expired dq (- i k))
            dq2 (pop-smaller-from-back dq1 nums (nth nums i))
            dq3 (conj dq2 i)
            result2 (if (>= i (dec k))
                      (conj result (nth nums (first dq3)))
                      result)]
        (recur (inc i) dq3 result2)))))
